californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -396,11 +398,12 @@ public void completed(final Exchange exchange) {
 
 		@Override
 		public void contextEstablished(final Exchange exchange) {
-
-			if (exchange.getRequest() != null) {
-				observationStore.setContext(exchange.getRequest().getToken(), exchange.getCorrelationContext());
+			Request currentRequest = exchange.getCurrentRequest(); 
+			Request request = exchange.getRequest(); 
+			if (request == currentRequest && request != null && request.isObserve()) {
+				observationStore.setContext(request.getToken(), exchange.getCorrelationContext());
```

Comment:
I used `XyzStorage` for both `ObservationStore `and `MessageExchangeStore`.

Currently, I've only saw implementation for observes (californium and leshan). 

https://github.com/eclipse/californium/issues/311#issuecomment-305957631

I think together "for all request" and "processing must be completed before returning" this is very dangerous for the performance. Therefore I would go for changing the setContext API to enable the implementation to select the way of processing the context (currently only filter for observes).